File: repl-samples/level4_linkedlist.py
# Linked List
# Implements a singly linked list with insert, search, and print
# Concepts: classes (structs), references (pointers), data structures

class Node:
    """A single node in the linked list"""
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    """Singly linked list with head pointer"""
    def __init__(self):
        self.head = None
        self.size = 0

    def push_front(self, value):
        """Insert a new node at the front"""
        node = Node(value)
        node.next = self.head
        self.head = node
        self.size += 1

    def find(self, value):
        """Search for a value, return True if found"""
        current = self.head
        while current is not None:
            if current.value == value:
                return True
            current = current.next
        return False

    def print_list(self):
        """Print all elements: [a -> b -> c]"""
        result = "["
        current = self.head
        while current is not None:
            if current != self.head:
                result += " -> "
            result += str(current.value)
            current = current.next
        result += "]"
        print(result)

# Main program
print("=== Linked List ===")
print("")

# Build a list by pushing to front
ll = LinkedList()
values = [10, 20, 30, 40, 50]
for v in values:
    ll.push_front(v)
    print("Pushed %d, size = %d" % (v, ll.size))

print("")
print("List contents:")
ll.print_list()

print("")

# Search for values
search_for = [30, 99, 10]
for v in search_for:
    if ll.find(v):
        print("Found %d in list" % v)
    else:
        print("%d not in list" % v)

File: repl-samples/level4_linkedlist.c
/* Linked List
   Implements a singly linked list with insert, search, and print
   Concepts: structs, pointers, memory allocation, data structures */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* A single node in the linked list */
typedef struct Node {
    int value;
    struct Node *next;
} Node;

/* The linked list with head pointer and size */
typedef struct {
    Node *head;
    int size;
} LinkedList;

/* Initialize an empty list */
void list_init(LinkedList *ll) {
    ll->head = NULL;
    ll->size = 0;
}

/* Insert a new node at the front */
void push_front(LinkedList *ll, int value) {
    Node *node = (Node *)malloc(sizeof(Node));
    node->value = value;
    node->next = ll->head;
    ll->head = node;
    ll->size++;
}

/* Search for a value, return 1 if found */
int find(LinkedList *ll, int value) {
    Node *current = ll->head;
    while (current != NULL) {
        if (current->value == value) {
            return 1;
        }
        current = current->next;
    }
    return 0;
}

/* Print all elements: [a -> b -> c] */
void print_list(LinkedList *ll) {
    char result[256] = "[";
    Node *current = ll->head;
    int first = 1;
    while (current != NULL) {
        if (!first) {
            strcat(result, " -> ");
        }
        char num[16];
        sprintf(num, "%d", current->value);
        strcat(result, num);
        first = 0;
        current = current->next;
    }
    strcat(result, "]");
    printf("%s\n", result);
}

/* Free all nodes in the list */
void list_free(LinkedList *ll) {
    Node *current = ll->head;
    while (current != NULL) {
        Node *temp = current;
        current = current->next;
        free(temp);
    }
    ll->head = NULL;
    ll->size = 0;
}

int main(void) {
    printf("=== Linked List ===\n");
    printf("\n");

    /* Build a list by pushing to front */
    LinkedList ll;
    list_init(&ll);

    int values[] = {10, 20, 30, 40, 50};
    for (int i = 0; i < 5; i++) {
        push_front(&ll, values[i]);
        printf("Pushed %d, size = %d\n", values[i], ll.size);
    }

    printf("\n");
    printf("List contents:\n");
    print_list(&ll);

    printf("\n");

    /* Search for values */
    int search_for[] = {30, 99, 10};
    for (int i = 0; i < 3; i++) {
        if (find(&ll, search_for[i])) {
            printf("Found %d in list\n", search_for[i]);
        } else {
            printf("%d not in list\n", search_for[i]);
        }
    }

    /* Clean up allocated memory */
    list_free(&ll);

    return 0;
}
